src/compiler/checker.ts

```diff
@@ -25284,6 +25284,55 @@ namespace ts {
                         }
                     }
                 }
+                if (type && type!==anyType && isPropertyAccessExpression(callExpression.expression)){
+                    // if (myDebug){
+                    //     consoleLog(`narrowTypeByCallExpression[dbg,in]: ==== `);
+                    // }
+                    const tmpSymbol = getSymbolOfNameOrPropertyAccessExpression(callExpression.expression);
+                    const declaration = tmpSymbol ? tmpSymbol.declarations? tmpSymbol.declarations[0] : undefined : undefined;
+                    const funcNode = declaration && (declaration as any).type
+                        && (declaration as any).type.kind === SyntaxKind.FunctionType
+                        ? (declaration as any).type as FunctionDeclaration : undefined;
+                    const funcRtnType = funcNode?.type ? getTypeOfNode(funcNode.type) : undefined;

```

Comment:
Notice that the function return type `funcRtnType` is not being narrowed by the function argument types.  
 - In case of no arguments (as in the original bug post) this is not an issue.   
 - In the case where there are arguments, then to the extent that the argument types could be resolved,  the  information could be used along with the signature information to narrow the return type `funcRtnType`.  However, IF the reason we got to this branch in the code is ONLY because   
   (a) there are no arguments and therefore a signature was not created, or   
   (b) we don't want to try to resolve due to expense, circularity, etc. 
then this is not an issue.  It is hard to tell from the code at the top two branches of `narrowTypeByCallExpression` (or elsewhere prior to calling this function) whether (a) || (b) is a correct assumption or not. This fix is ASSUMING that (a) || (b) is a valid assumption.